### DISCUSSION TITLE
Generics for UiState

### DISCUSSION THREAD

**Discussion by aartikov** [skydoves/sealedx#1] [General]
Hi, skydoves 🖐

Your library looks interesting, but I propose an alternative solution for `UiState`.

Let's make it generic:
```kotlin
sealed interface UiState<out T : Any> {
    data class Success<T : Any>(val data: T) : UiState<T>
    object Error : UiState<Nothing>
    object Loading : UiState<Nothing>
}
```

So, we can create `UiState.Success<Posters>(...)` or `UiState.Success<UserModel>(...)`. Also, it is not a problem to place a list inside - `UiState.Success<List<WhatsAppUser>>(...)`.

Generics allow to write generic code (sorry for the tautology 😅). For example, we can write an universal composable function to display `UiState`:
```kotlin
@Composable
fun <T : Any> LceUi(
    state: UiState<T>,
    modifier: Modifier = Modifier,
    content: @Composable (data: T) -> Unit
) {
    when (state) {
        is UiState.Success -> content(state.data)

        is UiState.Loading -> FullscreenCircularProgress(modifier)

        is UiState.Error -> ErrorPlaceholder(modifier)
    }
}
```

It is harder to make such reusable functions for a bunch of code-generated classes. Maybe, code generation is not a best solution here. Did I miss something? What do you think?

**Comment by skydoves** [comment]
Hey @aartikov, 

Good point! You can achieve the goals below with this library:
- Explicit class naming.
- Encapsulating the generic data types as a 'state' itself.
- Support more flexible extensions for each data type (working on).

But this is very experimental yet, so I would appreciate you if you could let me any issues or feedback. Thanks!

**Reply by skydoves** [reply]
Additionally, Your suggestion could be a great solution in cases of each screen has only one dedicated UI state, but I'm worried about the case where we have to have different UI states for the same model on the same screen.

More than anything, I hope we could find other use cases not only to stick to MVI cases but also other platforms, such as the backend side.
